File: src/lib/api/branch.ts
import { invoke } from '@tauri-apps/api/core';
import type { BranchInfo, FileDiffDetail, GitCredentials, SmartSyncResult, StashDetail, StashInfo, TagInfo } from '../types';
import { isTauri } from './client';
import { getMockBranches } from './mocks';

export async function getBranches(path: string): Promise<BranchInfo[]> {
  if (isTauri) {
    return await invoke<BranchInfo[]>('get_branches', { path });
  }
  return getMockBranches();
}

export async function deleteBranch(path: string, branchName: string, isRemote: boolean = false): Promise<void> {
  if (isTauri) {
    return await invoke<void>('delete_branch', { path, branchName, isRemote });
  }
}

export async function renameBranch(path: string, oldName: string, newName: string): Promise<void> {
  if (isTauri) {
    return await invoke<void>('rename_branch', { path, oldName, newName });
  }
}

export async function createBranch(
  path: string,
  newName: string,
  fromRef: string,
  checkout: boolean = true,
): Promise<void> {
  if (isTauri) {
    return await invoke<void>('create_branch', { path, newName, fromRef, checkout });
  }
}

export async function checkoutBranch(path: string, branchName: string): Promise<void> {
  if (isTauri) {
    return await invoke<void>('checkout_branch', { path, branchName });
  }
}

export async function getTags(path: string): Promise<TagInfo[]> {
  if (isTauri) {
    return await invoke<TagInfo[]>('get_tags', { path });
  }
  return [];
}

export async function getStashes(path: string): Promise<StashInfo[]> {
  if (isTauri) {
    return await invoke<StashInfo[]>('get_stashes', { path });
  }
  return [];
}

export async function stashSave(path: string, message: string, includeUntracked: boolean = true): Promise<string> {
  if (isTauri) {
    return await invoke<string>('stash_save', { path, message, includeUntracked });
  }
  return 'mock_stash_oid';
}

export async function stashApply(path: string, index: number): Promise<void> {
  if (isTauri) {
    await invoke('stash_apply', { path, index });
  }
}

export async function stashPop(path: string, index: number): Promise<void> {
  if (isTauri) {
    await invoke('stash_pop', { path, index });
  }
}

export async function stashDrop(path: string, index: number): Promise<void> {
  if (isTauri) {
    await invoke('stash_drop', { path, index });
  }
}

export async function getStashDetail(path: string, index: number): Promise<StashDetail> {
  if (isTauri) {
    return await invoke<StashDetail>('get_stash_detail', { path, index });
  }
  return {
    index,
    message: `WIP on main: mock stash ${index}`,
    commit_id: 'mock_stash_commit',
    branch_name: 'main',
    created_at: Math.floor(Date.now() / 1000) - 3600,
    files: [
      { path: 'src/App.svelte', status: 'modified', additions: 15, deletions: 4, is_untracked: false },
      { path: 'src/config.ts', status: 'added', additions: 8, deletions: 0, is_untracked: false },
    ],
    total_additions: 23,
    total_deletions: 4,
  };
}

export async function getStashFileDiff(
  path: string,
  index: number,
  filePath: string,
  ignoreWhitespace?: boolean
): Promise<FileDiffDetail> {
  if (isTauri) {
    return await invoke<FileDiffDetail>('get_stash_file_diff', {
      path,
      index,
      filePath,
      ignoreWhitespace,
    });
  }
  return {
    path: filePath,
    is_staged: false,
    is_binary: false,
    additions: 5,
    deletions: 2,
    hunks: [],
    original_content: '// Original code\nexport const val = 1;',
    modified_content: '// Modified in stash\nexport const val = 2;\nexport const extra = true;',
  };
}

export async function stashBranch(
  path: string,
  index: number,
  branchName: string
): Promise<BranchInfo> {
  if (isTauri) {
    return await invoke<BranchInfo>('stash_branch', { path, index, branchName });
  }
  return {
    name: `refs/heads/${branchName}`,
    shorthand: branchName,
    is_head: true,
    is_remote: false,
    ahead_count: 0,
    behind_count: 0,
    target_commit_id: 'mock_base_commit',
  };
}

export async function smartSync(
  path: string,
  remoteName?: string,
  branchName?: string,
  credentials?: GitCredentials
): Promise<SmartSyncResult> {
  if (isTauri) {
    return await invoke<SmartSyncResult>('smart_sync', {
      path,
      remoteName,
      branchName,
      credentials,
    });
  }
  return {
    branch_name: branchName || 'main',
    remote_name: remoteName || 'origin',
    status: 'Synced with remote successfully (Mock)',
  };
}

export async function getMergedBranches(
  path: string,
  baseBranch?: string
): Promise<string[]> {
  if (isTauri) {
    return await invoke<string[]>('get_merged_branches', { path, baseBranch });
  }
  return ['feature/old-dashboard', 'chore/cleanup-assets'];
}

export async function deleteMergedBranches(
  path: string,
  branchNames: string[]
): Promise<number> {
  if (isTauri) {
    return await invoke<number>('delete_merged_branches', { path, branchNames });
  }
  return branchNames.length;
}

File: src-tauri/src/commands/branch.rs
use tauri::{command, AppHandle, Emitter, State};
use crate::commands::state::AppState;
use crate::error::{AppError, AppResult};
use crate::git::{
    auth::GitCredentials,
    branches::{
        checkout_branch as git_checkout_branch, create_branch as git_create_branch,
        delete_branch as git_delete_branch, rename_branch as git_rename_branch,
        get_all_branches, get_all_stashes, get_all_tags,
    },
    commit_ops::{stash_apply as git_stash_apply, stash_drop as git_stash_drop, stash_pop as git_stash_pop, stash_save as git_stash_save},
    diff::FileDiffDetail,
    repo::open_repository as git_open_repo,
    stash_ops::{get_stash_detail as git_get_stash_detail, get_stash_file_diff as git_get_stash_file_diff, stash_branch as git_stash_branch, StashDetail},
    sync::{smart_sync_upstream_with_progress, SmartSyncResult, TransferProgressPayload},
    BranchInfo, StashInfo, TagInfo,
};

#[command]
pub async fn rename_branch(path: String, old_name: String, new_name: String) -> AppResult<()> {
    tokio::task::spawn_blocking(move || {
        let repo = git_open_repo(&path)?;
        git_rename_branch(&repo, &old_name, &new_name)
    })
    .await
    .map_err(|e| AppError::Internal(e.to_string()))?
}

#[command]
pub async fn get_branches(path: String) -> AppResult<Vec<BranchInfo>> {
    tokio::task::spawn_blocking(move || {
        let repo = git_open_repo(&path)?;
        get_all_branches(&repo)
    })
    .await
    .map_err(|e| AppError::Internal(e.to_string()))?
}

#[command]
pub async fn delete_branch(path: String, branch_name: String, is_remote: bool) -> AppResult<()> {
    tokio::task::spawn_blocking(move || {
        let repo = git_open_repo(&path)?;
        git_delete_branch(&repo, &branch_name, is_remote)
    })
    .await
    .map_err(|e| AppError::Internal(e.to_string()))?
}

#[command]
pub async fn create_branch(
    path: String,
    new_name: String,
    from_ref: String,
    checkout: bool,
) -> AppResult<()> {
    tokio::task::spawn_blocking(move || {
        let repo = git_open_repo(&path)?;
        git_create_branch(&repo, &new_name, &from_ref, checkout)
    })
    .await
    .map_err(|e| AppError::Internal(e.to_string()))?
}

#[command]
pub async fn checkout_branch(path: String, branch_name: String) -> AppResult<()> {
    tokio::task::spawn_blocking(move || {
        let repo = git_open_repo(&path)?;
        git_checkout_branch(&repo, &branch_name)
    })
    .await
    .map_err(|e| AppError::Internal(e.to_string()))?
}

#[command]
pub async fn get_tags(path: String) -> AppResult<Vec<TagInfo>> {
    tokio::task::spawn_blocking(move || {
        let repo = git_open_repo(&path)?;
        get_all_tags(&repo)
    })
    .await
    .map_err(|e| AppError::Internal(e.to_string()))?
}

#[command]
pub async fn get_stashes(path: String) -> AppResult<Vec<StashInfo>> {
    tokio::task::spawn_blocking(move || {
        let mut repo = git_open_repo(&path)?;
        get_all_stashes(&mut repo)
    })
    .await
    .map_err(|e| AppError::Internal(e.to_string()))?
}

#[command]
pub async fn stash_save(
    path: String,
    message: String,
    include_untracked: bool,
) -> AppResult<String> {
    tokio::task::spawn_blocking(move || {
        let mut repo = git_open_repo(&path)?;
        git_stash_save(&mut repo, &message, include_untracked)
    })
    .await
    .map_err(|e| AppError::Internal(e.to_string()))?
}

#[command]
pub async fn stash_apply(path: String, index: usize) -> AppResult<()> {
    tokio::task::spawn_blocking(move || {
        let mut repo = git_open_repo(&path)?;
        git_stash_apply(&mut repo, index)
    })
    .await
    .map_err(|e| AppError::Internal(e.to_string()))?
}

#[command]
pub async fn stash_pop(path: String, index: usize) -> AppResult<()> {
    tokio::task::spawn_blocking(move || {
        let mut repo = git_open_repo(&path)?;
        git_stash_pop(&mut repo, index)
    })
    .await
    .map_err(|e| AppError::Internal(e.to_string()))?
}

#[command]
pub async fn stash_drop(path: String, index: usize) -> AppResult<()> {
    tokio::task::spawn_blocking(move || {
        let mut repo = git_open_repo(&path)?;
        git_stash_drop(&mut repo, index)
    })
    .await
    .map_err(|e| AppError::Internal(e.to_string()))?
}

#[command]
pub async fn get_stash_detail(path: String, index: usize) -> AppResult<StashDetail> {
    tokio::task::spawn_blocking(move || {
        let mut repo = git_open_repo(&path)?;
        git_get_stash_detail(&mut repo, index)
    })
    .await
    .map_err(|e| AppError::Internal(e.to_string()))?
}

#[command]
pub async fn get_stash_file_diff(
    path: String,
    index: usize,
    file_path: String,
    ignore_whitespace: Option<bool>,
) -> AppResult<FileDiffDetail> {
    tokio::task::spawn_blocking(move || {
        let mut repo = git_open_repo(&path)?;
        git_get_stash_file_diff(&mut repo, index, &file_path, ignore_whitespace)
    })
    .await
    .map_err(|e| AppError::Internal(e.to_string()))?
}

#[command]
pub async fn stash_branch(path: String, index: usize, branch_name: String) -> AppResult<BranchInfo> {
    tokio::task::spawn_blocking(move || {
        let mut repo = git_open_repo(&path)?;
        git_stash_branch(&mut repo, index, &branch_name)
    })
    .await
    .map_err(|e| AppError::Internal(e.to_string()))?
}

#[command]
pub async fn smart_sync(
    path: String,
    remote_name: Option<String>,
    branch_name: Option<String>,
    credentials: Option<GitCredentials>,
    app_handle: AppHandle,
    state: State<'_, AppState>,
) -> AppResult<SmartSyncResult> {
    let effective_creds = match credentials {
        Some(c) => Some(c),
        None => {
            if let Ok(Some(acc)) = state.account_store.get_active_account(None) {
                Some(GitCredentials {
                    auth_type: "https_token".to_string(),
                    ssh_passphrase: None,
                    username: Some(acc.username),
                    token: Some(acc.token),
                })
            } else {
                None
            }
        }
    };
    tokio::task::spawn_blocking(move || {
        let repo = git_open_repo(&path)?;
        let app = app_handle.clone();
        smart_sync_upstream_with_progress(
            &repo,
            remote_name.as_deref(),
            branch_name.as_deref(),
            effective_creds,
            Some(move |progress: TransferProgressPayload| {
                let _ = app.emit("remote://transfer-progress", &progress);
            }),
        )
    })
    .await
    .map_err(|e| AppError::Internal(e.to_string()))?
}

#[command]
pub async fn get_merged_branches(path: String, base_branch: Option<String>) -> AppResult<Vec<String>> {
    tokio::task::spawn_blocking(move || {
        let repo = git_open_repo(&path)?;
        crate::git::branches::get_merged_branches(&repo, base_branch.as_deref())
    })
    .await
    .map_err(|e| AppError::Internal(e.to_string()))?
}

#[command]
pub async fn delete_merged_branches(path: String, branch_names: Vec<String>) -> AppResult<usize> {
    tokio::task::spawn_blocking(move || {
        let repo = git_open_repo(&path)?;
        crate::git::branches::delete_merged_branches(&repo, &branch_names)
    })
    .await
    .map_err(|e| AppError::Internal(e.to_string()))?
}

